src/core/service.c

```diff
@@ -1450,20 +1460,26 @@ static bool service_shall_restart(Service *s) {
 
 static void service_enter_dead(Service *s, ServiceResult f, bool allow_restart) {
         int r;
+        bool shall_restart;
+
         assert(s);
 
         if (s->result == SERVICE_SUCCESS)
                 s->result = f;
 
+        shall_restart = allow_restart && service_shall_restart(s);
+        if (shall_restart)
+                /* Do not close stored fds in the call to service_set_state below */
+                s->preserve_fds = true;
+
         service_set_state(s, s->result != SERVICE_SUCCESS ? SERVICE_FAILED : SERVICE_DEAD);
```

Comment:
Hmmm, I think we probably should skip going through SERVICE_FAILED/SERVICE_DEAD altogether if we are going to restart the unit anyway. Instead we should go directly to SERVICE_AUTO_RESTART. That appears more correct to me, and solves the issue in a prettier way.